File: 20.有效的括号.java
/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');
        Stack<Character> stask = new Stack<>();
        for(int i = 0; i < s.length(); i ++){
            char c = s.charAt(i);
            if(map.containsKey(c)){
                char e = stask.empty() ? '#' : stask.pop();
                if(e != map.get(c)){
                    return false;
                }
            }else{
                stask.push(c);
            }
        }
        
        return stask.empty();
    }
}
// @lc code=end

File: 20.有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
func isValid(s string) bool {
	var m = map[byte]byte{
		')': '(',
		']': '[',
		'}': '{',
	}
	var stack = make([]byte, 0)
	for i := 0; i < len(s); i++ {
		c := s[i]
		if rev, ok := m[c]; ok {
			if len(stack) == 0 {
				return false
			}
			sc := stack[len(stack)-1]
			if sc != rev {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, c)
		}
	}
	return len(stack) == 0
}

// @lc code=end
